synapse/handlers/device.py

```diff
@@ -505,6 +506,85 @@ async def user_left_room(self, user, room_id):
             await self.store.mark_remote_user_device_list_as_unsubscribed(user_id)
 
+    async def store_dehydrated_device(
+        self,
+        user_id: str,
+        device_data: JsonDict,
+        initial_device_display_name: Optional[str] = None,
+    ) -> str:
+        """Store a dehydrated device for a user.  If the user had a previous
+        dehydrated device, it is removed.
+
+        Args:
+            user_id: the user that we are storing the device for
+            device_data: the dehydrated device information
+            initial_device_display_name: The display name to use for the device
+        Returns:
+            device id of the dehydrated device
+        """
+        device_id = await self.check_device_registered(
+            user_id, None, initial_device_display_name,
+        )
+        old_device_id = await self.store.store_dehydrated_device(
+            user_id, device_id, device_data
+        )
+        if old_device_id is not None:
+            await self.delete_device(user_id, old_device_id)
+        return device_id
+
+    async def get_dehydrated_device(
+        self, user_id: str
+    ) -> Optional[Tuple[str, JsonDict]]:
+        """Retrieve the information for a dehydrated device.
+
+        Args:
+            user_id: the user whose dehydrated device we are looking for
+        Returns:
+            a tuple whose first item is the device ID, and the second item is
+            the dehydrated device information
+        """
+        return await self.store.get_dehydrated_device(user_id)
+
+    async def rehydrate_device(
+        self, user_id: str, access_token: str, device_id: str
+    ) -> dict:
+        """Process a rehydration request from the user.
+
+        Args:
+            user_id: the user who is rehydrating the device
+            access_token: the access token used for the request
+            device_id: the ID of the device that will be rehydrated
+        Returns:
+            a dict containing {"success": True}
+        """
+        success = await self.store.remove_dehydrated_device(user_id, device_id)
+
+        if success:
+            # If the dehydrated device was successfully deleted (the device ID
+            # matched the stored dehydrated device), then modify the access
+            # token to use the dehydrated device's ID and copy the old device
+            # display name to the dehydrated device, and destroy the old device
+            # ID
+            old_device_id = await self.store.set_device_for_access_token(
+                access_token, device_id
+            )
+            old_device = await self.store.get_device(user_id, old_device_id)
+            await self.store.update_device(
+                user_id, device_id, old_device["display_name"]
+            )
+            await self.store.delete_device(user_id, old_device_id)
+            await self.store.delete_e2e_keys_by_device(
+                user_id=user_id, device_id=old_device_id
+            )
```

Comment:
Can you add a comment above this saying why we aren't calling `self.delete_device` here? I *think* it is because we want to avoid the call to `delete_access_tokens_for_user`, but I'm not 100% sure.